pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,271 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/ReentrancyGuard.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
+
+import "../BaseWeightedPool.sol";
+import "./WeightCompression.sol";
+
+/**
+ * @dev Weighted Pool with mutable weights, designed to support V2 Liquidity Bootstrapping
+ */
+contract InvestmentPool is BaseWeightedPool, ReentrancyGuard {
+    // The Pause Window and Buffer Period are timestamp-based: they should not be relied upon for sub-minute accuracy.
+    // solhint-disable not-rely-on-time
+
+    using FixedPoint for uint256;
+    using WordCodec for bytes32;
+    using WeightCompression for uint256;
+
+    // State variables
+
+    // The current number of tokens in the pool
+    // Technically redundant; cached here to avoid calling getTokens on the pool,
+    //   which would be very gas-intensive for large numbers of tokens
+    // Can only change if tokens are added/removed
+    uint256 private _totalTokens;
+
+    // Store scaling factor and start/end weights for each token
+    // Mapping should be more efficient than trying to compress it further
+    // into a fixed array of bytes32 or something like that, especially
+    // since tokens can be added/removed - and re-ordered in the process
+    // For each token, we store:
+    // [ 155 bits|   5 bits |  32 bits   |   64 bits    |
```

Comment:
Why is the start weight at double precision compared to the end weight? In the LBP, this was necessary because there was basically just enough room to pack everything this way. But here, it seems like the end weight could just as easily be 64 bits, or each weight could even be 100+ bits; there is plenty of room. Are we saving it for something else?

Reply:
Yes - there is a lot more to store later (circuit breaker stuff, mainly). If necessary, the start weight precision could be reduced later (32 is probably still ok).

It is still desirable to have start weights higher precision than end weights. End weights are set by the user, presumably to "round numbers" - while start weights can be anything (e.g., if they restart the weight update in the middle). Higher resolution start weights help avoid (potentially game-able) weight "jumps" due to lack of precision.

Reply:
See my comment above re. allowing for smaller weights - they will essentially be fixed for pools with a large amount of tokens unless we reduce the minimum one.